Emit every n seconds

albertteoh opened the pull request:
(Resubmitting PR as per: https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/17063#issuecomment-1353869053)

**Description:** 
Fixes the `value not found in metricKeyToDimensions` error that appears after the metrics key cache is full.

This is achieved by emitting all metrics with cached dimension keys at configurable intervals (defaulting to `15s`) as suggested by @bogdandrutu in [this comment](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/16311#pullrequestreview-1188519807). 

This also aims to address a problem reported by users where CPU utilisation gradually increases over time, particularly with high cardinality dimensions. See: https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/9018#issuecomment-1252476417.

As the metrics are emitted at regular intervals, a separate goroutine is dedicated to out-of-band building and emitting of metrics to downstream metrics consumers; metrics aggregation is still performed in-band. 

Prior to this PR, if the dimension keys are not in cache (removed from eviction list), the `value not found in metricKeyToDimensions` would be returned. What also exacerbated this problem was that accesses to evicted items did not revive those keys back into the LRU cache.

**Link to tracking Issue:** 
Fixes #9018
Fixes #15688

Continues from #16024, #16311 and #16601

**Testing:** 

- Bug was easily reproducible by reducing the configurable dimensions key cache size down to a small value e.g. `2`.
- Wrote a unit test that reproduces the bug and implement the fix to pass the test.
- This has been exercised in a production environment with real data and confirmed it appears to have resolved the high CPU utilisation problem.

albertteoh committed:
WIP

albertteoh committed:
Publish metrics only if key is cached

albertteoh committed:
Emit metrics every n period

albertteoh committed:
Update tests

albertteoh committed:
Gofumpt

albertteoh committed:
Fix tests

albertteoh committed:
Remove unused context

albertteoh committed:
Remove unused context

albertteoh committed:
Fix test and better comments

albertteoh committed:
Better shutdown testing

albertteoh committed:
Merge branch 'main' into emit-every-n-seconds

albertteoh committed:
Fix import groupings

albertteoh committed:
Remove time.Sleep where possible

albertteoh committed:
behaviour -> behavior

albertteoh committed:
Merge branch 'main' into emit-every-n-seconds

albertteoh committed:
Merge branch 'main' into emit-every-n-seconds

albertteoh committed:
Merge branch 'main' into emit-every-n-seconds

albertteoh committed:
Merge branch 'main' into emit-every-n-seconds

albertteoh committed:
go mod tidy

albertteoh committed:
Shutdown only once

Signed-off-by: albertteoh <see.kwang.teoh@gmail.com>

albertteoh committed:
Better name

Signed-off-by: albertteoh <see.kwang.teoh@gmail.com>

albertteoh committed:
Breaking change

Signed-off-by: albertteoh <see.kwang.teoh@gmail.com>

albertteoh committed:
No need for p variable

Signed-off-by: albertteoh <see.kwang.teoh@gmail.com>

albertteoh committed:
Merge branch 'main' into emit-every-n-seconds

albertteoh committed:
Merge branch 'main' into emit-every-n-seconds

albertteoh committed:
Fix ordering of sleep and clock tick

albertteoh committed:
Remove invalid TestShutdown due to sync.Once

albertteoh committed:
Prevent ConsumeTraces and ConsumeMetrics race

srikanthccv commented on processor/spanmetricsprocessor/processor.go:
```diff
@@ -308,6 +328,12 @@ func (p *processorImp) collectLatencyMetrics(ilm pmetric.ScopeMetrics) error {
 	for key, hist := range p.histograms {
+		// If the key is not in the cache, simply don't publish it because it's likely
+		// the metric has expired (or the cache size is too small).
+		dimensions, ok := p.metricKeyToDimensions.Get(key)
+		if !ok {
+			continue
+		}
```
Isn't this silently ignoring the error? Wouldn't we still want to log the error message that the key was not found? Otherwise, how would one know why some metrics are missing from the collection?

albertteoh replied:
Yup, good point, thanks for pointing this out. I've added a warning log message and removed the ineffectual comment because I think the log message should document the cause of the issue well enough. Done in: [8c66291](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/17307/commits/8c662917005db89ca868339d841fbaf88e52bb3c).

albertteoh committed:
Use struct{} instead of bool

Signed-off-by: albertteoh <see.kwang.teoh@gmail.com>

albertteoh committed:
Add warn log message

Signed-off-by: albertteoh <see.kwang.teoh@gmail.com>

albertteoh commented on processor/spanmetricsprocessor/processor.go:
```diff
@@ -85,6 +83,12 @@ type processorImp struct {
 	metricKeyToDimensions *cache.Cache[metricKey, pcommon.Map]
+
+	ticker  *clock.Ticker
+	done    chan bool
```
Yup, thanks, done in: [44699f3](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/17307/commits/44699f3dbac1d10abc632c2ec4fbf483a1834edf).

albertteoh commented:
I noticed the [failing tests](https://github.com/open-telemetry/opentelemetry-collector-contrib/actions/runs/3819937914/jobs/6497928788#step:9:7275) were from `TestParser/RFC3164`, and doesn't appear to be caused by this PR.

albertteoh committed:
Merge branch 'main' into emit-every-n-seconds

albertteoh committed:
Merge branch 'main' into emit-every-n-seconds

albertteoh committed:
Remove unused dependencies

Signed-off-by: Albert Teoh <see.kwang.teoh@gmail.com>

MovieStoreGuy commented on processor/spanmetricsprocessor/factory.go:
```diff
@@ -44,9 +46,10 @@ func createDefaultConfig() component.Config {
 }
 
 func createTracesProcessor(_ context.Context, params processor.CreateSettings, cfg component.Config, nextConsumer consumer.Traces) (processor.Traces, error) {
-	return newProcessor(params.Logger, cfg, nextConsumer)
+	return newProcessor(params.Logger, cfg, nextConsumer, clock.Realtime().NewTicker(cfg.(*Config).MetricsFlushInterval))
```
To allow for using this within testing, you can use:

```suggestion
func createTracesProcessor(ctx context.Context, params processor.CreateSettings, cfg component.Config, nextConsumer consumer.Traces) (processor.Traces, error) {
	return newProcessor(params.Logger, cfg, nextConsumer, clock.FromContext(ctx).NewTicker(cfg.(*Config).MetricsFlushInterval))
```

albertteoh replied:
Good suggestion, done in: [75ddc2b](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/17307/commits/75ddc2bfbbdd81f14716fa6b1adf9f62c7b06031)

MovieStoreGuy commented on processor/spanmetricsprocessor/processor.go:
```diff
@@ -260,31 +290,21 @@ func (p *processorImp) tracesToMetrics(ctx context.Context, traces ptrace.Traces
-	if err = p.metricsExporter.ConsumeMetrics(ctx, m); err != nil {
-		return err
+	if err := p.metricsExporter.ConsumeMetrics(ctx, m); err != nil {
+		p.logger.Error("Failed ConsumeMetrics: " + err.Error())
```
Use structured logs to make it easier to search in logging platforms:

```suggestion
		p.logger.Error("Failed ConsumeMetrics", zap.Error(err))
```

albertteoh replied:
Thanks, addressed in: [261ebbb](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/17307/commits/261ebbb816f031a9f4cfdd5a432a87a7f40a53e9)

MovieStoreGuy reviewed (commented):
Happy new year! 🎆 

Only couple comments from me.

albertteoh committed:
Use clock.FromContext

albertteoh committed:
Use structured logging

Signed-off-by: albertteoh <see.kwang.teoh@gmail.com>

albertteoh committed:
Merge branch 'emit-every-n-seconds' of github.com:albertteoh/opentelemetry-collector-contrib into emit-every-n-seconds

albertteoh commented:
> Happy new year! 🎆
> 
> Only couple comments from me.

Happy new year to you too, @MovieStoreGuy! Thanks for the review, I have addressed your comments.

albertteoh committed:
Update test

Signed-off-by: albertteoh <see.kwang.teoh@gmail.com>

albertteoh committed:
Merge branch 'main' into emit-every-n-seconds

albertteoh committed:
Merge branch 'main' into emit-every-n-seconds

albertteoh committed:
go mod tidy

albertteoh commented:
Hi 👋🏼 , friendly nudge to see if an approver could please take a look? Thanks!

MovieStoreGuy reviewed (approved):
These changes make sense to me, and I am not too concerned with the breaking change since this component is highlighted as `InDevelopment` so there is warnings that are shown on collector start up that it can change.
